overlord/state/state.go

```diff
@@ -72,35 +72,48 @@ func (data customData) set(key string, value interface{}) {
 // The state is persisted on every unlock operation via the StateBackend
 // it was initialized with.
 type State struct {
-	// locking
 	mu  sync.Mutex
 	muC int32
-	// storage
+
 	backend Backend
 	data    customData
 	changes map[string]*Change
 	tasks   map[string]*Task
+
+	modified bool
 }
 
 // New returns a new empty state.
 func New(backend Backend) *State {
 	return &State{
-		backend: backend,
-		data:    make(customData),
-		changes: make(map[string]*Change),
-		tasks:   make(map[string]*Task),
+		backend:  backend,
+		data:     make(customData),
+		changes:  make(map[string]*Change),
+		tasks:    make(map[string]*Task),
+		modified: true,
 	}
 }
 
+// Modified returns whether the state was modified since the last checkpoint.
+func (s *State) Modified() bool {
+	return s.modified
+}
+
 // Lock acquires the state lock.
 func (s *State) Lock() {
 	s.mu.Lock()
 	atomic.AddInt32(&s.muC, 1)
 }
 
-func (s *State) ensureLocked() {
-	c := atomic.LoadInt32(&s.muC)
-	if c != 1 {
+func (s *State) reading() {
+	if atomic.LoadInt32(&s.muC) != 1 {
+		panic("internal error: accessing state without lock")
+	}
+}
+
+func (s *State) writing() {
+	s.modified = true
```

Comment:
That's what I did too, until I realized testing was simpler this way. Since the check actually blows up, there's no actual benefit in doing it afterwards, so simpler testing won.